documentation/projects/proposals/popularity_optimizations/20230420-implementation_plan_popularity_optimizations.md

```diff
@@ -0,0 +1,488 @@
+popularity refresh task will take a considerably long time to complete.
+
+To avoid this, we must break up the popularity refresh into smaller
+transactions, committing after every 10,000 rows are updated (this number can be
+tweaked as necessary). This ensures that a concurrently running provider DAG
+will only ever have to wait for a second or two before locked rows are released.
+
+The approach is easiest explained with some sample code:
+
+```pgsql
+-- The final product will be generalized by provider and media type.
+-- In a local test, this successfully updated ~1 million Flickr
+-- records in ~6 seconds, while the reingestion DAG was running.
+DO $$
+DECLARE
+  rows_per_loop int := 10000;
+  min_row int := 1;
+  max_row int;
+  max_timestamp timestamp;
+BEGIN
+  -- We will use this to select rows that have not been recently updated.
+  SELECT NOW() INTO max_timestamp;
+
+  -- A temporary table storing the identifiers of all records that need
+  -- to be updated, plus their row number. The row number will be used to
+  -- paginate the results more efficiently than using OFFSET.
+  CREATE TEMP TABLE rows_to_update AS
+    SELECT ROW_NUMBER() over() row_id, identifier
+    FROM image
+    WHERE provider='flickr' AND updated_on < max_timestamp;
+  CREATE INDEX ON rows_to_update(row_id);
+
+  SELECT COUNT(*) INTO max_row FROM rows_to_update;
+
+  -- Loop over `rows_to_update` 10k rows at a time, updating scores.
+  -- Avoid deadlock using SKIP LOCKED
+  FOR i IN min_row..max_row BY rows_per_loop LOOP
+   UPDATE image
+   SET standardized_popularity = standardized_image_popularity(
+       image.provider,
+       image.meta_data
+   )
+   WHERE identifier in (
+       SELECT identifier FROM rows_to_update
+       WHERE row_id >= i AND row_id < i+rows_per_loop
+       FOR UPDATE SKIP LOCKED
+   );
+   -- COMMIT this batch
+   COMMIT;
+END LOOP;
+END;
+$$;
+```
+
+This is much less efficient than doing a single `UPDATE`, but has clear
+advantages:
+
+- It allows provider DAGs to continue running unimpeded.
+- The data refresh can run at any time, even while the refresh tasks are
+  running.
+- Because updates are committed incrementally, each batch of updated scores will
+  be 'live' on the image table as soon as they are committed. Rather than
+  waiting for all rows to be calculated before making the update, as much
+  up-to-date data as possible will be present each time the data refresh runs.
+
+### Remove popularity steps from the data_refresh DAGs
+
+We can simply remove the popularity tasks from the relevant data refresh DAG
+factories. See the `Parallelizable Streams` section for more information on the
+timing of this change.
+
+### Run the data refresh from the media tables
+
+Once the `image` and `audio` tables contain up-to-date standardized popularity
+data, we can update the data refresh to read from these tables rather than their
+respective materialized views. Because the schema of the media tables now
```

Comment:
Now we use the materialized views for selecting data during the data refresh.
Would it have any effect on the ingestion or data refresh performance if we use `image` tables both for refreshing data and for ongoing writing of the data from provider API scripts? My guess is no, considering that we do use the `image` table for creating the `materialized view` anyways.